temporal-sdk/src/main/java/io/temporal/internal/worker/TrackingSlotSupplier.java

```diff
@@ -0,0 +1,132 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.internal.worker;
+
+import com.uber.m3.tally.NoopScope;
+import com.uber.m3.tally.Scope;
+import io.temporal.worker.MetricsType;
+import io.temporal.worker.tuning.SlotPermit;
+import io.temporal.worker.tuning.SlotReleaseReason;
+import io.temporal.worker.tuning.SlotReservationContext;
+import io.temporal.worker.tuning.SlotSupplier;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Wraps a slot supplier and supplements it with additional tracking information that is useful to
+ * provide to all implementations. This type is used internally rather than {@link SlotSupplier}
+ * directly.
+ *
+ * @param <SI> The slot info type
+ */
+public class TrackingSlotSupplier<SI> {
+  private final SlotSupplier<SI> inner;
+  private final AtomicInteger issuedSlots = new AtomicInteger();
+  private final Map<SlotPermit, SI> usedSlots = new ConcurrentHashMap<>();
+  private Scope metricsScope;
+
+  public TrackingSlotSupplier(SlotSupplier<SI> inner) {
+    this.inner = inner;
+    metricsScope = new NoopScope();
+  }
+
+  public SlotPermit reserveSlot(SlotReservationData dat) throws InterruptedException {
+    SlotPermit p = inner.reserveSlot(createCtx(dat));
+    issuedSlots.incrementAndGet();
+    return p;
+  }
+
+  public Optional<SlotPermit> tryReserveSlot(SlotReservationData dat) {
+    Optional<SlotPermit> p = inner.tryReserveSlot(createCtx(dat));
+    if (p.isPresent()) {
+      issuedSlots.incrementAndGet();
+    }
+    return p;
+  }
+
+  public void markSlotUsed(SI slotInfo, SlotPermit permit) {
+    if (permit == null) {
+      return;
+    }
+    inner.markSlotUsed(slotInfo, permit);
+    usedSlots.put(permit, slotInfo);
+    publishSlotsMetric();
+  }
+
+  public void releaseSlot(SlotReleaseReason reason, SlotPermit permit) {
+    if (permit == null) {
+      return;
+    }
+    inner.releaseSlot(reason, permit);
+    issuedSlots.decrementAndGet();
+    usedSlots.remove(permit);
+    publishSlotsMetric();
+  }
+
+  public int maximumSlots() {
+    return inner.maximumSlots();
+  }
+
+  public int getIssuedSlots() {
+    return issuedSlots.get();
+  }
+
+  public void setMetricsScope(Scope metricsScope) {
+    this.metricsScope = metricsScope;
+  }
+
+  Map<SlotPermit, SI> getUsedSlots() {
+    return usedSlots;
+  }
+
+  private void publishSlotsMetric() {
+    this.metricsScope
+        .gauge(MetricsType.WORKER_TASK_SLOTS_AVAILABLE)
+        .update(maximumSlots() - usedSlots.size());
+  }
+
+  private SlotReservationContext<SI> createCtx(SlotReservationData dat) {
+    return new SlotResCtx(dat.getTaskQueue(), Collections.unmodifiableMap(usedSlots));
+  }
+
+  private class SlotResCtx implements SlotReservationContext<SI> {
```

Comment:
Java doesn't usually abbreviate like this (even for private stuff despite the annoyance to the author)